Dev doc guide (#13163)
openvinotoolkit/openvino: docs/dev/dev_doc_guide.md

# Creating developer documentation

Developer documentation provides an overview of the component development and helps engineers understand the underlying technology.
This guide contains the requirements for writing developer documentation.

## The component structure

> The OpenVINO component structure is fully defined in the [dedicated section of developer documentation](./index.md#openvino-component-structure),

Author: ```suggestion
The OpenVINO component structure is fully defined in the [dedicated section of developer documentation](./index.md#openvino-component-structure).
```

> The OpenVINO component structure is fully defined in the [dedicated section of developer documentation](./index.md#openvino-component-structure),
> this guide is more concentrated on parts which are used for developer documentation, the structure you can find below:

Reviewer: ```suggestion
The OpenVINO component structure is fully defined in the [dedicated section of developer documentation](./index.md#openvino-component-structure),
This guide highlights the parts which are used for developer documentation. The structure is presented below:
```

Reviewer: @ilyachur just one minor grammar correction, otherwise lgtm

# How to develop developer documentation

The developer documentation is needed to provide basic information about component development.
This guide contains requirements which are applied for component developer documentation.

## The component structure

The OpenVINO component structure is fully defined in the [dedicated section of developer documentation](./index.md#openvino-component-structure),
this guide is more concentrated on parts which are used for developer documentation, the structure you can find below:

```
openvino_component/
    docs/               // Contains detailed component documentation
    README.md           // Endtry point for the developer documentation
```

> 

Author: Here we need to provide information only about some parts of the component

> 

Reviewer: I would actually like to ask what do we mean by component exactly? Because according to this, the documentation is a separate OV component? Is that right?

> 

Author: In general the component it is a separate part of OpenVINO toolkit. It can be user documentation, sample, transformations, frontend, plugin and etc.
Each component can contain different parts:
```
ov_component/           // Component folder
    cmake/              // (optional) CMake scripts which are related only for this component
    dev_api/            // (optional) Developer API is used in case if component wants to provide some API for internal developers
    docs/               // (optional) Contains detailed component documentation
    include/            // (optional) Public component API
    src/                // Sources of the component
    tests/              // Tests on the component
    thirdparty/         // (optional) Third-party dependencies
    CMakeLists.txt      // Main CMake script
    README.md           // (optional) Endtry point for the developer documentation
```

Developer documentation is a part of component.

if we are talking about user documentation from the developer perspective it is also a separate component which is located in the `docs/` folder. Also we have a separate cmake target for this component

## ***README.md*** content

README.md should provide an entry point to component documentation. 
First of all the readme shouldn't be overloaded by redundant information, has a strict user friendly structure and provide references to additional documentation.

The proposed template for Readme file is:
```md
# Name of component

Short information about the component.
 * Responsibility
 * Supported features
 * Code style

> ## Key persons

Reviewer: ```suggestion
## Key contacts
```

Section provides informations about groups of developers who can help in case of questions and also review and merge PRs.

## Components

Section contains common infromation about included components: API, sources, tests and etc.
Detailed information can be located in the `docs/` folder, this section can contain links to these pages.

## Architecture

Is an optional charpter which provide main ideas about the component architecture.
It can contains references to dedicated pages with additional information.

## Tutorials

Section contains a list of component tutorials.
Example:
 * [How to support new operation](./docs/operation_support.md)
 * [How to implement new feature](./docs/new_feature.md)
 * [How to debug the component](./docs/debug_capabilities.md)
 * ...

## See also

Section contains a list of related pages.
Example:
 * [OpenVINO™ README](../../README.md)
 * [Developer documentation](../../docs/dev/index.md)
 * ...
```

## ***docs/*** folder content

> Docs folder should contain media content for the developer documentation and dedicated pages with detailed information about component features.
> Pages in this folder also shouldn't be overloaded by information, each page should provide information only which is related with original topic. If we need to provide some additional information it should be a reference to the separate file which user can open and get all necessary information about this additional topic.

Reviewer: ```suggestion
Docs folder should contain pages with a detailed description of the component features and media content for the developer documentation. Make sure that the pages in this folder provide information directly related to the topic. It is better to provide information about additional topics in a separate file and make a reference to it.
```

> ## Approaches for the documentation development
> 
> As already was said before the developer documentation should be clear and has a strong structure. Each page should provide only information which relates to the main topic, in case if we need to provide an additional information, it will be better to have a dedicated page.
> 
> In order to decrease the complexity the developer documentation and make it more user-friendly, it can contain a lot of different images, diagrams, examples which should illustrate described things.
> 
> For diagrams we are using `mermaid` framework to get more information about it please read [Github Documentation](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-diagrams). The framework allows to create diagrams and past it in your documents, for example the code:

Reviewer: ```suggestion
## Media content
The developer documentation may include pictures, diagrams, and examples that illustrate the described features to make it easier to understand. 
It is recommended to use the `mermaid` framework for diagrams. Read the [Github Documentation](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-diagrams) to get more information. The framework allows creating diagrams and pasting them to your documents. For example, the code:
```
